### ISSUE TITLE
Within-batch negative sampling for multiplicative models

### ISSUE THREAD

**Issue by apoorvumang** [uma-pi1/kge#127]
Hello

This is just a suggestion for a feature that I'll explain below:

I have been working on applying KG embeddings to question answering, and for this I required fast embedding learning on large KGs (millions of entities and triples). To do this efficiently on a single GPU, KvsAll is definitely not feasible and I used Neg Sampling. The normal implementation of negsampling requires some fixed n number of negative samples for each positive sample(triple) in the batch size b. However I think there is a more efficient way (which I used):

1. Use the grouping method as used in KvsAll: for each (h,r) pair, prepare a list of positive tails (ids of these entities). 
2. Initialize a entity list tail_list= []. For each (h,r) pair in batch, keep appending the positive tails list to tail_list. This will serve as a proxy to the list of all entities for this batch. Clearly, this will be much less than the order of millions.
3. To prepare the target vector for each (h,r) in the batch, set all indices corresponding to its positive tails in tail_list to 1, others to 0
4. Use the normal scoring function/loss function as used in KvsAll, except at the point where we take dot product with the full entity embeddings list, we use tail_list entity embeddings list.

This way we are using all other elements in batch as negative samples. Optionally one can also add some more random negative samples as well (as done below). This enables batch sizes of upto 1000, along with faster scoring of each batch, which results in significant speedup for me.

Here is the code I use for generating the batch (might not make sense without the rest of the code):

```
def get_batch(self, er_vocab, er_vocab_pairs, idx, batch_size, neg_batch_size=10000, eval=False):
        batch = er_vocab_pairs[idx:idx+batch_size]
        total_pos_length = 0
        for idx, pair in enumerate(batch):
            positives = er_vocab[pair]
            total_pos_length += len(positives)
        neg_batch_size = neg_batch_size - total_pos_length
        if neg_batch_size < 0:
            neg_batch_size = 0
        targets = torch.zeros([len(batch), neg_batch_size + total_pos_length], dtype=torch.float32)
        tail_list = []
        if self.cuda:
            targets = targets.cuda()
        start_id = 0    
        for idx, pair in enumerate(batch):
            positives = er_vocab[pair]
            tail_list += positives
            for i in range(len(positives)):
                targets[idx, i + start_id] = 1.
            start_id += len(positives)
        negatives = np.random.randint(len(d.entities), size=neg_batch_size)
        tail_list = np.concatenate((np.array(tail_list),negatives))
        return np.array(batch), tail_list, targets
```

In my experiments, this technique performed well on small datasets as well (FB15k-237).

I will be happy to share the full code for this technique, if it makes sense to you.

Thanks
Apoorv

**Comment by samuelbroscheit** [comment]
Hi Apoorv,

for my experiments in https://www.aclweb.org/anthology/2020.acl-main.209.pdf I did a similar thing (see page 12). By doing this construction of pseudo negative examples you'll end up sampling an entity approx. proportional to its frequency in the training data, which I also found to work well, but not 100% as good as full 1vsAll/KvsAll on small datasets (but this was not rigorously investigated). 

Here in LibKGE we have shared negative sampling and also reusing the scores in the batch see https://github.com/uma-pi1/kge/blob/d918c48dbbc9448bdea7edba182906505f59aca2/kge/config-default.yaml#L282 and https://github.com/uma-pi1/kge/blob/d918c48dbbc9448bdea7edba182906505f59aca2/kge/config-default.yaml#L294 for documentation on that. With those settings we could train on Wikidata-5m (see https://github.com/uma-pi1/kge#wikidata5m-wikidata ) and converge within a few hours @AdrianKs please chime in, if there is more to elaborate on this.

**Comment by apoorvumang** [comment]
I have used dropout and batch normalization as was done in TuckER, which can be the reason for different results. Let me benchmark this on Wikidata-5m and I'll get back to you with the results (performance + time taken)

Thanks!

**Comment by apoorvumang** [comment]
Will also check out training with shared negative sampling using LibKGE on my KG (Freebase subset) and see the results

**Comment by AdrianKs** [comment]
Also when training with negative sampling on large graphs make sure to use sparse embeddings and weighted regularization. This should reduce memory consumption and allow quicker training.

**Comment by apoorvumang** [comment]
> Also when training with negative sampling on large graphs make sure to use sparse embeddings and weighted regularization. This should reduce memory consumption and allow quicker training.

Can you elaborate on sparse embeddings? Do you have any documentation about it? Thanks

**Comment by samuelbroscheit** [comment]
Currently the main source of documentation are the comments in the config files, in this case https://github.com/uma-pi1/kge/blob/master/kge/model/embedder/lookup_embedder.yaml . Sparse embeddings refers to the "sparse" option of PyTorch's Embedding layer, which will result in having gradients only for the rows of the embedding matrix that have been used in the batch. Similarly, the weighted regularization only computes a weight penalty for the embeddings used in the batch, and thus will also result in having gradients only for those.

**Comment by apoorvumang** [comment]
Got it, thanks! Closing this issue since I got all my answers

**Comment by rgemulla** [comment]
Perhaps we should create a Wiki and collect this kind of information. @samuelbroscheit Perhaps you can start and copy your comments above to the wiki.
